src/integrations/pulsar/consumer.cpp

```diff
@@ -54,14 +55,16 @@ utils::BasicResult<std::string, std::vector<Message>> GetBatch(TConsumer &consum
   auto remaining_timeout_in_ms = info.batch_interval.value_or(kDefaultBatchInterval).count();
   auto start = std::chrono::steady_clock::now();
 
-  for (int64_t i = 0; remaining_timeout_in_ms > 0 && i < batch_size && is_running.load(); ++i) {
+  while (remaining_timeout_in_ms > 0 && batch.size() < batch_size && is_running.load()) {
     pulsar_client::Message message;
     const auto result = ConsumeMessage(consumer, message, remaining_timeout_in_ms);
     switch (result) {
       case pulsar_client::Result::ResultTimeout:
         return std::move(batch);
       case pulsar_client::Result::ResultOk:
-        batch.emplace_back(Message{std::move(message)});
+        if (message.getMessageId() != last_message_id) {
+          batch.emplace_back(Message{std::move(message)});
+        }
```

Comment:
Because the docs say that it should give the next message after the one you set but for some reason, it doesn't. I checked so many times the logs of Pulsar on client and server side and it doesn't make sense...
It starts from the message you send to the reader so this is a way of skipping it. Maybe it's just an edge case that happens in E2E tests but for now, I would leave it as safety.